include/execution.hpp

```diff
@@ -2171,7 +2172,257 @@ namespace std::execution {
+      };
+
+    template <class _SenderId, class _ReceiverId>
+      class __operation : public __operation_base {
+        using _Sender = __t<_SenderId>;
+        using _Receiver = __t<_ReceiverId>;
+
+        struct __on_stop_requested {
+          in_place_stop_source& __stop_source_;
+          void operator()() noexcept {
+            __stop_source_.request_stop();
+          }
+        };
+        using __on_stop = optional<typename stop_token_of_t<
+            env_of_t<_Receiver> &>::template callback_type<__on_stop_requested>>;
+
+        _Receiver __recvr_;
+        __on_stop __on_stop_{};
+        shared_ptr<__sh_state<_Sender>> __shared_state_;
+
+        template <class... _As>
+          requires tag_invocable<set_value_t, _Receiver, decay_t<_As>&...>
+        void __apply_values(tuple<_As...>& __tupl) noexcept {
+          try {
+            std::apply([&](auto&... __args) -> void {
+              execution::set_value((_Receiver&&) __recvr_, __args...);
+            }, __tupl);
+          } catch(...) {
+            execution::set_error((_Receiver&&) __recvr_, current_exception());
+          }
+        }
+
+        template <class... _As>
+          requires (!tag_invocable<set_value_t, _Receiver, decay_t<_As>&...>)
+        void __apply_values(tuple<_As...>&) noexcept {
+          assert(!"_Should never get here");
+        }
+
+        void __propagate_signal() noexcept {
+          auto &__data = __shared_state_->__data_;
+
+          if (__data.index() == 0) {
+            visit([&](auto& __tupl) -> void {
+              __apply_values(__tupl);
+            }, std::get<0>(__data));
+          }
+          if (__data.index() == 1) {
+            visit([&](auto& __err) -> void {
+              execution::set_error((_Receiver&&) __recvr_, __err);
+            }, std::get<1>(__data));
+          }
+          else if (__data.index() == 2) {
+            execution::set_stopped((_Receiver&&) __recvr_);
+          }
+        }
+
+      public:
+        __operation(_Receiver&& __rcvr,
+                    shared_ptr<__sh_state<_Sender>> __shared_state)
+          : __recvr_((_Receiver&&)__rcvr)
+          , __shared_state_(move(__shared_state)) {
+        }
+
+        void __notify() noexcept override {
+          __on_stop_.reset();
+          __propagate_signal();
+        }
+
+        friend void tag_invoke(start_t, __operation& __self) noexcept try {
+          __sh_state<_Sender> *__shared_state = __self.__shared_state_.get();
+          unique_lock __lock{__shared_state->__mutex_};
+          __state_t __state = __shared_state->__state_;
+
+          if (__state == __state_t::__completed) {
+            __lock.unlock();
+            __self.__propagate_signal();
+          }
+          else {
+            __self.__on_stop_.emplace(
+                get_stop_token(get_env(__self.__recvr_)),
+                __on_stop_requested{__shared_state->__stop_source_});
+
+            if (__shared_state->__stop_source_.stop_requested()) {
```

Comment:
@lewissbaker can you pls weigh in on whether this registration of the stop callback is safe? I think it is, but I wanted to be sure.